### PR TITLE
fix for the case when fields are unmodifiable collection

### PR DESCRIPTION
**Pull request by izebit**
Hi. 
I found bug. It happens when object has  unmodifiable collection field. 

For example
```java
class A <T>{
     Collection<T>collection = Collections.emptyList();
}
```
Somewhere in program ... 
```java
A<String> obj = new A<>();
obj = Arrays.asList("hello","world");
```

Maybe, i missed place where need fix yet.

### CODE REVIEW

File: protostuff-collectionschema/src/main/java/io/protostuff/MapSchema.java [hunk 1]
```diff
@@ -47,6 +47,44 @@
...
+    public static <T extends Map> T removeUnmodifiableWrapper(T map, MessageFactory messageFactory)
+    {
+        if(map == null)
+            return null;
+
+        try
+        {
+           map.put(EMPTY_OBJECT, EMPTY_OBJECT);
+           map.remove(EMPTY_OBJECT);
+        }
```
Comment (dyu) [inline protostuff-collectionschema/src/main/java/io/protostuff/MapSchema.java]: Not in favor of this hack.
Surely there is a solution to know if a collection is unmodifable without adding/removing?

Comment (izebit) [inline protostuff-collectionschema/src/main/java/io/protostuff/MapSchema.java]: Yes of cource, but the solution, in my opinion, is more robust than other solutions. 

I saw several ways to determine a unmodifiable collection:

1. Add/remove element and catch `UnsupportOperationException`
2. Compare types, for example:
```java 
List<String> list = new ArrayList<>();
List<String> unmodifiableList = Collections.unmodifiableList(list);
boolean isUnmodifiable = Collections.unmodifiableList(list).getClass()
                                                            .isInstance(unmodifiableList); 
```
3. It is the continuation of the second solution. Create list of types and compare with them.
```java  
for (String className: classNames)
       if (collection.getClass().getSimpleName().equals(className))
	      return true;
```
The second and third solutions is more elegant than first, but they are not flexible. We create list with unmodifiable collection types from `java.util.Collections.*` and we don't respect for any unmodifiable colections from other libraries.

I will glad, if you offer other solution for this problem.

Comment (dyu) [inline protostuff-collectionschema/src/main/java/io/protostuff/MapSchema.java]: Can you try setting this property and see if you still get the failures?
```
-Dprotostuff.runtime.always_use_sun_reflection_factory=true

// or via System.setProperty on your main class
```

A better solution is to mark the field with:
```
@Tag(value = 1, merge = false) // to be added.
Collection<String> foo = ...; // any immutable init you can think of
```
That way, the deserializer would always overwrite the field with a new one.

Comment (izebit) [inline protostuff-collectionschema/src/main/java/io/protostuff/MapSchema.java]: It is very interesting. The errors are not happens when i set this property. 

I agree with you. The solution with annotation is more pretty then what i suggested. I will try to implement this feature.

File: protostuff-runtime/src/main/java/io/protostuff/runtime/RuntimeCollectionFieldFactory.java [hunk 2]
```diff
@@ -506,4 +505,20 @@ public FieldType getFieldType()
...
+    @SuppressWarnings("unchecked")
+    private static<T extends Collection<?>, E extends Enum<E>> void merge(Accessor accessor,
+                                                                          Object message,
+                                                                          Input input,
+                                                                          Schema<T> schema,
+                                                                          MessageFactory messageFactory,
+                                                                          Class<E>... enumType) throws IOException
+    {
+        T collection = accessor.get(message);
+        if (enumType.length == 0)
```
Comment (kshchepanovskyi) [inline protostuff-runtime/src/main/java/io/protostuff/runtime/RuntimeCollectionFieldFactory.java]: This change will introduce a significant performance penalty, as for every element of collection you call `removeUnmodifiableWrapper`. Probably there is better solution, like catch `UnsupportedOperationException` when trying to add element to a collection field and creating new collection instance in that case.

**Comment by kshchepanovskyi** [comment]
This change will introduce a significant performance penalty, as for every element of collection that is merged into a message you call `removeUnmodifiableWrapper`. Probably there is better solution, like catch `UnsupportedOperationException` when trying to add element to a collection field and creating new collection instance in that case.

**Pull request closed without merging**